### PR TITLE
Integration of CalcDeltaImpact class of unsequa into develop

### PR DESCRIPTION
**Pull request by simonameiler**
Changes proposed in this PR:
- Integration of CalcDeltaImpact class of unsequa into develop
- Update unsequa tutorial
- Modification of plotting functionalities of the uncertainty output in `unc_output` to skip columns of the uncertainty output dataframe which are zeros or nan.
- addition of a small function called `save_divide` in `util` which handles division by zero and NaN values in the numerator or denominator. This is needed for the relative impact change calculation (delta impact). Tests for the function are also added.

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]
- [x] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

**Comment by chahank** [comment]
Can you please merge the develop branch into this branch? All tests are now failing, but I think it is due to a bug that is fixed on develop.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+from climada.util import log_level
+
+LOGGER = logging.getLogger(__name__)
+
+class CalcDeltaImpact(Calc):
+    """
+    Impact uncertainty caclulation class.
+
+    This is the class to perform uncertainty analysis on the outputs of a
+    climada.engine.impact.Impact() object.
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:53]: I think this could be updated.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Define Uncertainty Impact class
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:19]:
```suggestion
Define Uncertainty class for Impact differences between two climates.
```

File: climada/engine/unsequa/calc_delta_climate.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+    ):
+        """Initialize UncCalcImpact
+
+        Sets the uncertainty input variables, the impact metric_names, and the
+        units.
+
+        Parameters
+        ----------
+        exp_initial_input_var : climada.engine.uncertainty.input_var.InputVar or climada.entity.Exposure
+            Exposure uncertainty variable or Exposure
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:115]: I think the docstrings year need a bit more clarity. It is not simply any Exposure. In the end, reported uncertainties will be final - initial impacts, which should be made clear here too. idem for haz and impfset.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+                    unc_sample,
+                    rp=None,
+                    calc_eai_exp=False,
+                    calc_at_event=False,
+                    processes=1,
+                    chunksize=None
+                    ):
+        """
+        Computes the differential impact between the reference (initial) and 
+        future (final)for each sample in unc_data.sample_df.
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:151]:
```suggestion
        future (final) for each sample in unc_data.sample_df.
```

File: climada/engine/unsequa/calc_delta_climate.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+
+        By default, the aggregated average impact within a period of 1/frequency_unit
+        (impact.aai_agg) and the excees impact at return periods rp
+        (imppact.calc_freq_curve(self.rp).impact) is computed.
+        Optionally, eai_exp and at_event is computed (this may require
+        a larger amount of memory if the number of samples and/or the number
+        of centroids and/or exposures points is large).
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:158]: Maybe here one should explain a bit better what exactly are the values reported. Is it the difference of the outputs such as `eai_exp`, or is the difference first done on the impact matrix and then the outputs are computed? This is in particular would make a difference for the return periods.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+            at_event_list
+        ] =  self._compute_imp_metrics(
+            samples_df, chunksize=chunksize, processes=processes
+            )
+
+        # Assign computed impact distribution data to self
+        aai_agg_unc_df  = pd.DataFrame(aai_agg_list,
+                                                columns = ['aai_agg'])
+        freq_curve_unc_df = pd.DataFrame(freq_curve_list,
+                                    columns=['rp' + str(n) for n in rp])
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:257]: I think the formating is slightly off here.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+
+def _safe_divide(numerator, denominator, replace_with=np.nan):
+    """
+    Safely divide two arrays or scalars, handling division by zero and NaN values.
+
+    Parameters
+    ----------
+    numerator (array-like or scalar): Numerator for division.
+    denominator (array-like or scalar): Denominator for division.
+    replace_with (float): Value to use in place of division by zero or NaN. Defaults to NaN.
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:333]: Could be changed to numpy doc style.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+
+    Returns
+    -------
+    array-like or scalar: Result of safe division.
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:337]: idem.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+                imp_metrics = itertools.starmap(
+                    _map_impact_calc, p_iterator
+                    )
+
+        #Perform the actual computation
+        with log_level(level='ERROR', name_prefix='climada'):
+            return _transpose_chunked_data(imp_metrics)
+
+
+def _safe_divide(numerator, denominator, replace_with=np.nan):
```
Comment (chahank) [inline climada/engine/unsequa/calc_delta_climate.py:325]: I would consider putting this method into the utils and importing it here. I think this does not belong to this module in particular.

File: climada/engine/unsequa/input_var.py [hunk 10]
```diff
@@ -68,40 +68,39 @@ class InputVar():
...
-    >>> import scipy as sp
-    >>> def litpop_cat(m, n):
-    ...     exp = Litpop.from_countries('CHE', exponent=[m, n])
-    ...     return exp
-    >>> distr_dict = {
-    ...     'm': sp.stats.randint(low=0, high=5),
-    ...     'n': sp.stats.randint(low=0, high=5)
-    ...     }
-    >>> iv_cat = InputVar(func=litpop_cat, distr_dict=distr_dict)
+        import scipy as sp
```
Comment (chahank) [inline climada/engine/unsequa/input_var.py:71]: This file did not need any modification. Why was the docstring style changed?

File: climada/engine/unsequa/unc_output.py [hunk 11]
```diff
@@ -424,9 +424,8 @@ def plot_sample(self, figsize=None):
         figsize : tuple(int or float, int or float), optional
             The figsize argument of matplotlib.pyplot.subplots()
             The default is derived from the total number of plots (nplots) as:
-
-            >>> nrows, ncols = int(np.ceil(nplots / 3)), min(nplots, 3)
-            >>> figsize = (ncols * FIG_W, nrows * FIG_H)
+                nrows, ncols = int(np.ceil(nplots / 3)), min(nplots, 3)
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:427]: Why the change in style?

File: climada/engine/unsequa/unc_output.py [hunk 12]
```diff
@@ -557,6 +556,14 @@ def plot_uncertainty(self, metric_list=None, orig_list=None, figsize=None,
...
                     ax.remove()
                 continue
+            # Check if the column data is empty or contains only NaNs
+            data, m_unit = u_vtm(unc_df_plt[col])
+            data = pd.Series(data)
+            if data.empty or data.isna().all() or data.dropna().shape[0] < 2:
+                print(f"Skipping plot for '{col}': insufficient data.")
+                if ax is not None:
+                    ax.remove()
+                continue
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:566]: Is this part of this PR? This seems not related to the delta climate uncetainty.

File: climada/engine/unsequa/unc_output.py [hunk 13]
```diff
@@ -659,7 +666,11 @@ def plot_rp_uncertainty(self, orig_list=None, figsize=(16, 6), axes=None):
         prop_cycle = plt.rcParams['axes.prop_cycle']
         colors = prop_cycle.by_key()['color']
 
+        # Check for NaNs in each column of unc_df and plot only if data is valid
         for n, (_name, values) in enumerate(unc_df.items()):
+            if values.isna().all() or len(values.dropna()) < 2:
+                print(f"Skipping plot for '{_name}': insufficient data.")
+                continue
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:673]: idem

File: climada/engine/unsequa/unc_output.py [hunk 14]
```diff
@@ -741,9 +752,8 @@ def plot_sensitivity(self, salib_si='S1', salib_si_conf='S1_conf',
         figsize : tuple(int or float, int or float), optional
             The figsize argument of matplotlib.pyplot.subplots()
             The default is derived from the total number of plots (nplots) as:
-
-            >>> nrows, ncols = int(np.ceil(nplots / 3)), min(nplots, 3)
-            >>> figsize = (ncols * FIG_W, nrows * FIG_H)
+                nrows, ncols = int(np.ceil(nplots / 3)), min(nplots, 3)
+                figsize = (ncols * FIG_W, nrows * FIG_H)
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:756]: Style?

File: climada/engine/unsequa/unc_output.py [hunk 15]
```diff
@@ -793,16 +803,11 @@ def plot_sensitivity(self, salib_si='S1', salib_si_conf='S1_conf',
 
         for ax, metric in zip(flat_axes, metric_list):
             df_S = self.get_sensitivity(salib_si, [metric]).select_dtypes('number')
-            if not df_S.columns[df_S.isnull().all()].empty:
-                LOGGER.warning("All-NaN columns encountered: %s",
-                               list(df_S.columns[df_S.isnull().all()]))
-            df_S = df_S.loc[:, df_S.notnull().any()]
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:799]: Reason?

File: climada/engine/unsequa/unc_output.py [hunk 16]
```diff
@@ -793,16 +803,11 @@ def plot_sensitivity(self, salib_si='S1', salib_si_conf='S1_conf',
...
             if df_S.empty:
                 ax.set_xlabel('Input parameter')
                 ax.remove()
                 continue
             df_S_conf = self.get_sensitivity(salib_si_conf, [metric]).select_dtypes('number')
-            df_S_conf = df_S_conf.loc[:, df_S.columns]
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:805]: Reason?

File: climada/engine/unsequa/unc_output.py [hunk 17]
```diff
@@ -845,9 +850,8 @@ def plot_sensitivity_second_order(self, salib_si='S2', salib_si_conf='S2_conf',
         figsize : tuple(int or float, int or float), optional
             The figsize argument of matplotlib.pyplot.subplots()
             The default is derived from the total number of plots (nplots) as:
-
-            >>> nrows, ncols = int(np.ceil(nplots / 3)), min(nplots, 3)
-            >>> figsize = (ncols * 5, nrows * 5)
+                nrows, ncols = int(np.ceil(nplots / 3)), min(nplots, 3)
+                figsize = (ncols * 5, nrows * 5)
```
Comment (chahank) [inline climada/engine/unsequa/unc_output.py:854]: Style?

**Comment by chahank** [comment]
I have one larger general question: from the naming, and what the docstring says, I would expect the new module to compute the absolute differences. However, the current code seems to compute relative differences. What is the rational here? This should maybe be an option? 

After these points are resolved, someone should still do a careful check of the actual methods. I briefly went over them, and it seems all fine. But details can be important, in particular also because the test is quite minimal.

File: climada/engine/unsequa/input_var.py (hunk 10 above)
Comment (simonameiler) [reply climada/engine/unsequa/input_var.py:71]: no idea. I reverted the change by replacing the file with the current version of input_var.py on develop.

File: climada/engine/unsequa/unc_output.py (hunk 12 above)
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:566]: It is. I slightly modified the plotting functionalities of the uncertainty output to skip columns of the uncertainty output dataframe which are zeros or nan.

I added this change to the description of the PR.

File: climada/engine/unsequa/unc_output.py (hunk 13 above)
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:673]: see comment above.

File: climada/engine/unsequa/unc_output.py (hunk 11 above)
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:427]: no idea. I reverted the change in style.

File: climada/engine/unsequa/unc_output.py (hunk 14 above)
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:756]: I don't know why parts of the code changed that I didn't mean to change. I reverted these parts back to what is in develop. This answer also applies to the next comments.

File: climada/engine/unsequa/test/test_unsequa.py [hunk 18]
```diff
@@ -327,6 +328,48 @@ def test_save_load_pass(self):
...
+    def test_calc_uncertainty_pass(self):
+        """Test compute the uncertainty distribution for an impact"""
+
+        exp_unc, impf_unc, _ = make_input_vars()
+        haz = haz_dem()
+        haz2 = haz_dem()
+        haz2.intensity *=2
+        unc_calc = CalcDeltaImpact(exp_unc, impf_dem(), haz, exp_dem(), impf_unc, haz2)
+        unc_data = unc_calc.make_sample(N=1)
+        unc_data = unc_calc.uncertainty(unc_data, calc_eai_exp=False, calc_at_event=False)
```
[the remark this answers is not in the document]
Comment (simonameiler) [reply climada/engine/unsequa/test/test_unsequa.py:343]: Okay, I changed the test to N=2. 

And added a test for the sensitivity default setup - with N=4 samples.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 1 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py:53]: to what?

File: climada/engine/unsequa/calc_delta_climate.py (hunk 5 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py:158]: okay, I added a specific description incl. example.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 7 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py:338]: okay. changed.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 9 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py:330]: done. the method is in util now. including unit tests. I added this aspect to the PR description too.

**Comment by simonameiler** [comment]
> I have one larger general question: from the naming, and what the docstring says, I would expect the new module to compute the absolute differences. However, the current code seems to compute relative differences. What is the rational here? This should maybe be an option?
> 
> After these points are resolved, someone should still do a careful check of the actual methods. I briefly went over them, and it seems all fine. But details can be important, in particular also because the test is quite minimal.

I resolved all points to my best knowledge. I don't think there should be an option to calculate delta impacts for absolute values. Absolute delta impact calculations are not useful without knowledge of a reference state. I added a few lines in the docstrings reasoning for the relative delta impact calculation.

File: climada/util/value_representation.py [hunk 19]
```diff
@@ -179,3 +177,66 @@ def value_to_monetary_unit(values, n_sig_dig=None, abbreviations=None):
...
+
+    >>> _safe_divide(1, 0, replace_with=0)
+    0
+
+    >>> _safe_divide([1, 0, 3], [0, 0, 3])
+    array([nan, nan,  1.])
+    """
+    # Handle array inputs
+    if not np.isscalar(numerator) or not np.isscalar(denominator):
+        # If either input is an array and contains NaN, replace the whole result with replace_with
```
Comment (emanuel-schmid) [inline climada/util/value_representation.py:224]: seems counter intuitive to me. why not `safe_divide([nan, nan, 1.], 4) == [nan, nan, 0.25]` ?
I don't understand why division by zero is dealt with element wise but if any of the elements is `nan` all should turn to `nan`?
Since `safe_divide(4, 0)` is equal to `safe_divide(4, nan)`, i.m.o., `safe_divide([4, 4], [1, 0])` should also be equal to `safe_divide([4, 4], [1, nan])`.

File: climada/util/value_representation.py [hunk 20]
```diff
@@ -179,3 +177,66 @@ def value_to_monetary_unit(values, n_sig_dig=None, abbreviations=None):
...
+
+    This function handles division by zero and NaN values in the numerator or denominator.
+    If the division results in infinity or NaN, it is replaced by a specified value.
+
+    Parameters
+    ----------
+    numerator : array-like or scalar
+        The numerator for division.
+    denominator : array-like or scalar
+        The denominator for division. Division by zero is handled safely.
```
Comment (emanuel-schmid) [inline climada/util/value_representation.py:194]: there are only tests for scalar and np.array, but it could be a just a list - right?

File: climada/util/value_representation.py [hunk 21]
```diff
@@ -179,3 +177,66 @@ def value_to_monetary_unit(values, n_sig_dig=None, abbreviations=None):
...
+    replace_with : float, optional
+        The value to use in place of division results that are infinity or NaN.
+        By default, it is NaN.
+
+    Returns
+    -------
+    array-like or scalar
```
Comment (emanuel-schmid) [inline climada/util/value_representation.py:201]: the return type for array-like input is `np.ndarray`

File: climada/util/value_representation.py (hunk 20 above)
Comment (simonameiler) [reply climada/util/value_representation.py]: yes, indeed. I added a test for lists too.

File: climada/util/value_representation.py (hunk 19 above)
Comment (simonameiler) [reply climada/util/value_representation.py]: good point. the choice in the safe_divide function to treat NaN values in array inputs differently than division by zero is indeed not necessary. I updated the function:

`    with np.errstate(divide="ignore", invalid="ignore"):
        result = np.true_divide(numerator, denominator)
        
        # Check if the result is a scalar
        if np.isscalar(result):
            if not np.isfinite(result):
                return replace_with
        else:
            # Replace infinities, NaNs, and division by zeros in np.ndarray
            result[~np.isfinite(result)] = replace_with

    return result`

File: climada/util/value_representation.py [hunk 22]
```diff
@@ -179,3 +177,66 @@ def value_to_monetary_unit(values, n_sig_dig=None, abbreviations=None):
...
+    array([nan, nan,  1.])
+    """
+    # Handle array inputs
+    if not np.isscalar(numerator) or not np.isscalar(denominator):
+        # If either input is an array and contains NaN, replace the whole result with replace_with
+        if np.any(np.isnan(numerator)) or np.any(np.isnan(denominator)):
+            if np.isscalar(numerator) or np.isscalar(denominator):
+                return replace_with
+            else:
+                return np.full(numerator.shape, replace_with)
```
[the remark this answers is not in the document]
Comment (simonameiler) [reply climada/util/value_representation.py]: With the revised safe_divide function, the specific code section is no longer present. However, I still added tests for NaN handling.

File: climada/engine/unsequa/unc_output.py [hunk 23]
```diff
@@ -1139,7 +1152,55 @@ def __init__(self, samples_df, unit, aai_agg_unc_df, freq_curve_unc_df, eai_exp_
...
+            Each row contains the values of eai_exp for one sample (row of
+            samples_df)
+        at_event_unc_df : pandas.DataFrame
+            Each row contains the values of at_event for one sample (row of
+            samples_df)
+        tot_value_unc_df : pandas.DataFrame
+            Each row contains the value of tot_value for one sample (row of
+            samples_df)
+        coord_df : pandas.DataFrame
+            Coordinates of the exposure
```
Comment (peanutfun) [inline climada/engine/unsequa/unc_output.py:1186]: Why not a `geopandas.GeoSeries`? Is there any expectation about the column names, like `latitude`, `longitude`? Could you add this to the documentation?

File: climada/engine/unsequa/calc_delta_climate.py [hunk 24]
```diff
@@ -0,0 +1,435 @@
...
+        "impf_initial_input_var",
+        "haz_initial_input_var",
+        "exp_final_input_var",
+        "impf_final_input_var",
+        "haz_final_input_var",
+    )
+    """Names of the required uncertainty variables"""
+
+    _metric_names = ("aai_agg", "freq_curve", "at_event", "eai_exp")
+    """Names of the cost benefit output metrics"""
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_delta_climate.py:99]: Why do these attributes exist? If I am not mistaken, they are never used anywhere in this class.

Comment (peanutfun) [reply climada/engine/unsequa/calc_delta_climate.py:99]: Ah, I see they are used in the base class. This is super confusing! In particular, the empty attributes of `Calc` with the same name are never overwritten here. @chahank, instead of relying of `self` to point to the derived class, it would probably be better to define `Calc.input_vars` as `@abstractmethod`, to clarify what needs to be overwritten by derived classes. See https://docs.python.org/3/library/abc.html#abc.abstractmethod

Comment (peanutfun) [reply climada/engine/unsequa/calc_delta_climate.py:99]: Alternative: Make `input_var_names` and `metric_names` parameters of the `Calc` base class and pass them through the base class constructor? This way, it would also be more explicit.

File: climada/engine/unsequa/calc_delta_climate.py [hunk 25]
```diff
@@ -0,0 +1,435 @@
...
+        exp_final_input_var : climada.engine.uncertainty.input_var.InputVar or climada.entity.Exposure
+            Exposure uncertainty variable or Exposure of final state
+        impf_final_input_var : climada.engine.uncertainty.input_var.InputVar or climada.entity.ImpactFuncSet
+            Impact function set uncertainty variable or Impact function set of final state
+        haz_final_input_var : climada.engine.uncertainty.input_var.InputVar or climada.hazard.Hazard
+            Hazard uncertainty variable or Hazard of final state
+
+        """
+
+        Calc.__init__(self)
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_delta_climate.py:132]: Not really necessary, but safer in general
```suggestion
        super().__init__(self)
```

File: climada/engine/unsequa/calc_delta_climate.py [hunk 26]
```diff
@@ -0,0 +1,435 @@
...
+        self.check_distr()
+
+    def uncertainty(
+        self,
+        unc_sample,
+        rp=None,
+        calc_eai_exp=False,
+        calc_at_event=False,
+        processes=1,
+        chunksize=None,
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_delta_climate.py:150]: If you have the time, could you add type hints?

File: climada/engine/unsequa/calc_delta_climate.py [hunk 27]
```diff
@@ -0,0 +1,435 @@
...
+
+
+def _map_impact_calc(
+    sample_chunks,
+    exp_initial_input_var,
+    impf_initial_input_var,
+    haz_initial_input_var,
+    exp_final_input_var,
+    impf_final_input_var,
+    haz_final_input_var,
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_delta_climate.py:334]: Rather trivial, but these parameters are missing from the docstring

File: climada/engine/unsequa/calc_delta_climate.py [hunk 28]
```diff
@@ -0,0 +1,435 @@
...
+            eai_exp_final = np.array([])
+
+        if calc_at_event:
+            at_event_initial = imp_initial.at_event
+            at_event_final = imp_final.at_event
+        else:
+            at_event_initial = np.array([])
+            at_event_final = np.array([])
+
+        delta_aai_agg = safe_divide(
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_delta_climate.py:411]: The `safe_divide` looks neat but it also seems a bit unnecessary to me. What's the problem with `inf` values? Why was it important to have NaNs exclusively?

File: climada/engine/unsequa/unc_output.py (hunk 12 above)
Comment (peanutfun) [reply climada/engine/unsequa/unc_output.py:567]: Good idea, but it looks a bit weird that the subplots are missing completely. Instead of removing the axes, you could make an empty plot and write a short text like "No data to plot"?

File: doc/tutorial/climada_engine_unsequa.ipynb [hunk 29]
```diff
@@ -2480,6 +2481,440 @@
...
+    "output_imp.get_samples_df().tail()\n",
+    "\n",
+    "output_imp = calc_imp.uncertainty(output_imp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Plotting functionalities work analogous to CalcImpact. However, the figure labelling is wrong. What we calculate and plot is the relative risk change in % and not an absolute number in USD!!!"
```
Comment (peanutfun) [inline doc/tutorial/climada_engine_unsequa.ipynb:2764]: Can this be fixed? 🤔

File: climada/engine/unsequa/calc_base.py [hunk 30]
```diff
@@ -323,21 +320,23 @@ def sensitivity(self, unc_output, sensitivity_method = 'sobol',
 
         Parameters
         ----------
-        unc_output : climada.engine.uncertainty.unc_output.UncOutput
+        unc_output : climada.engine.usequa.UncOutput
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_base.py:323]: Typo
```suggestion
        unc_output : climada.engine.unsequa.UncOutput
```

File: climada/engine/unsequa/calc_base.py [hunk 31]
```diff
@@ -323,21 +320,23 @@ def sensitivity(self, unc_output, sensitivity_method = 'sobol',
...
             https://salib.readthedocs.io/en/latest/api.html
-            Default: 'sobol'
         sensitivity_kwargs: dict, optional
             Keyword arguments of the chosen SALib analyse method.
             The default is to use SALib's default arguments.
 
         Returns
         -------
-        sens_output : climada.engine.uncertainty.unc_output.UncOutput()
+        sens_output : climada.engine.usequa.UncOutput
```
Comment (peanutfun) [inline climada/engine/unsequa/calc_base.py:339]:
```suggestion
        sens_output : climada.engine.unsequa.UncOutput
```

**Review by peanutfun — CHANGES_REQUESTED** [review]
This looks good and useful! There's not much left to do:
* Please resolve all linter issues (except `too-many-arguments` and `too-many-local-variables`). This should be rather simple
* The module documentation is missing. Please add an appropriate section to [`doc/climada/climada.engine.unsequa.rst`](https://raw.githubusercontent.com/CLIMADA-project/climada_python/develop/doc/climada/climada.engine.unsequa.rst) and double-check the rendered docs.
* The tutorial section on the new class should probably start with an explanation what the class does (current paragraphs 2 and 3), and then mention that it works analogously to the other class (paragraph 1)
* The table of contents in the tutorial was not updated, and the links don't work anyway. Maybe just remove it for now?

File: climada/engine/unsequa/calc_delta_climate.py (hunk 24 above)
Comment (chahank) [reply climada/engine/unsequa/calc_delta_climate.py:99]: @peanutfun : thank you for the insights on the code struture! All elements that one could consider, but I do not think that they are critical. I also think this PR is not the place for such discussion. The module is at the moment stable and functional. I would suggest to have a discussion on the needs of refactoring somewhere else.

File: climada/engine/unsequa/unc_output.py (hunk 23 above)
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:1186]: I just mirrored that from the `UncImpactOutput` class. You may need to consult with @chahank if you'd like to see such a change.

File: climada/engine/unsequa/unc_output.py [hunk 32]
```diff
@@ -1139,7 +1152,55 @@ def __init__(self, samples_df, unit, aai_agg_unc_df, freq_curve_unc_df, eai_exp_
...
+class UncDeltaImpactOutput(UncOutput):
+    """Extension of UncOutput specific for CalcDeltaImpact, returned by the  uncertainty() method.
+    """
+    def __init__(self, samples_df, unit, aai_agg_unc_df, freq_curve_unc_df, eai_exp_unc_df,
+                 at_event_initial_unc_df, at_event_final_unc_df, tot_value_unc_df, coord_df):
+        """Constructor
 
+        Uncertainty output values from impact.calc for each sample
+
+        Parameters
```
[the remark this answers is not in the document]
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:1164]: good catch. and there were tot_value parameters still in the docstring and as arguments which are not necessary. I updated the respective section.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 25 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py]: This actually created an error in the tests. I thus reverted it back to Calc.__init__(self).

File: climada/engine/unsequa/calc_delta_climate.py (hunk 24 above)
Comment (peanutfun) [reply climada/engine/unsequa/calc_delta_climate.py:99]: Absolutely agreed! My comments were a rather unordered stream of thoughts 🙈 I'll try to keep track of that in a separate Discussion, and resolve this conversation!

File: doc/tutorial/climada_engine_unsequa.ipynb (hunk 29 above)
Comment (simonameiler) [reply doc/tutorial/climada_engine_unsequa.ipynb:2764]: yes. I added a new argument (`calc_delta`) in the plotting functions, which allows you to print out x-axis labels correctly.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 28 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py:411]: In the context of the `_map_impact_calc` function, using `safe_divide` ensures that any calculations involving division are robust against undefined operations handling, statistical analysis integrity, data integrity preservation, library compatibility, and accurate visualization. The resulting dataset remains consistent, and further analysis or visualization steps are not misled by infinite or undefined values.

File: climada/engine/unsequa/unc_output.py (hunk 23 above)
Comment (simonameiler) [reply climada/engine/unsequa/unc_output.py:1186]: I'll let that up to you and @chahank to decide and suggest you modify the docstring of CalcImpact in the same go too. I don't know what the columns must be called.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 26 above)
Comment (simonameiler) [reply climada/engine/unsequa/calc_delta_climate.py:150]: here too, I mirrored `CalcImpact`. I give the ball back to you and @chahank for such modifications.

**Comment by simonameiler** [comment]
> This looks good and useful! There's not much left to do:
> 
> * Please resolve all linter issues (except `too-many-arguments` and `too-many-local-variables`). This should be rather simple
> * The module documentation is missing. Please add an appropriate section to [`doc/climada/climada.engine.unsequa.rst`](https://raw.githubusercontent.com/CLIMADA-project/climada_python/develop/doc/climada/climada.engine.unsequa.rst) and double-check the rendered docs.
> * The tutorial section on the new class should probably start with an explanation what the class does (current paragraphs 2 and 3), and then mention that it works analogously to the other class (paragraph 1)
> * The table of contents in the tutorial was not updated, and the links don't work anyway. Maybe just remove it for now?

I addressed all your points. In my eyes, it's good to go.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 28 above)
Comment (peanutfun) [reply climada/engine/unsequa/calc_delta_climate.py:419]: Excellent, thanks for the explanation!

File: climada/engine/unsequa/unc_output.py (hunk 23 above)
Comment (peanutfun) [reply climada/engine/unsequa/unc_output.py:1216]: Fine. Let's leave it as is.

File: climada/engine/unsequa/calc_delta_climate.py (hunk 26 above)
Comment (peanutfun) [reply climada/engine/unsequa/calc_delta_climate.py:158]: No worries. Let's leave it like that.

File: doc/climada/climada.engine.unsequa.rst [hunk 33]
```diff
@@ -41,3 +41,10 @@ climada\.engine\.unsequa\.unc\_output module
     :undoc-members:
     :show-inheritance:
 
+climada\.engine\.unsequa\.calc\_delta\_impact module
+---------------------------------------------
+
+.. automodule:: climada.engine.unsequa.calc_delta_impact
```
Comment (peanutfun) [inline doc/climada/climada.engine.unsequa.rst:47]:
```suggestion
climada\.engine\.unsequa\.calc\_delta\_climate module
---------------------------------------------

.. automodule:: climada.engine.unsequa.calc_delta_climate
```

**Review by peanutfun — APPROVED** [review]
Thank you @simonameiler! 🙌  I fixed the remaning whitespace issues and will merge as soon as the pipeline succeeds (which it should 🤞)
